src/mlpack/core/data/csv_parser.hpp

```diff
@@ -0,0 +1,110 @@
+/** 
+ * @file core/data/csv_parser.hpp
+ * @author Conrad Sanderson
+ * @author Gopi M. Tatiraju
+ *
+ * This csv parser is designed by taking reference from armadillo's csv parser.
+ * In this mlpack's version, all the arma dependencies were removed or replaced
+ * accordingly, making the parser totally independent of armadillo.
+ *
+ * This parser will be totally independent to any linear algebra library.
+ * This can be used to load data into any matrix, i.e. arma and bandicoot
+ * in future.
+ *
+ * https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.5.x/include/armadillo_bits/diskio_meat.hpp
+ * Copyright 2008-2016 Conrad Sanderson (http://conradsanderson.id.au)
+ * Copyright 2008-2016 National ICT Australia (NICTA)
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_CSV_PARSER_HPP
+#define MLPACK_CORE_DATA_CSV_PARSER_HPP
+
+#include <limits>
+
+namespace mlpack{
+namespace data{
+enum struct file_type : unsigned int
+{
+  mlp_file_type_unknown,
+  mlp_auto_detect,        //!< attempt to automatically detect the file type
+  mlp_raw_ascii,          //!< raw text (ASCII), without a header
+  mlp_arma_ascii,         //!< Armadillo text format, with a header specifying matrix type and size
+  mlp_csv_ascii,          //!< comma separated values (CSV), without a header
+  mlp_raw_binary,         //!< raw binary format (machine dependent), without a header
+  mlp_arma_binary,        //!< Armadillo binary format (machine dependent), with a header specifying matrix type and size
+  mlp_pgm_binary,         //!< Portable Grey Map (greyscale image)
+  mlp_ppm_binary,         //!< Portable Pixel Map (colour image), used by the field and cube classes
+  mlp_hdf5_binary,        //!< HDF5: open binary format, not specific to Armadillo, which can store arbitrary data
+  mlp_hdf5_binary_trans,  //!< [DO NOT US deprecated] as per hdf5_binary, but save/load the data with columns transposed to rows
+  mlp_coord_ascii         //!< simple co-ordinate format for sparse matrices (indices start at zero)
+  };
```

Comment:
These one need to be in an independent header file in order to decouple it from the parser since they are unrelated